File: windows/BambuBar.Windows/Services/MoonrakerStatusParser.cs
using System.Globalization;
using System.IO;
using System.Text.Json;
using BambuBar.Models;

namespace BambuBar.Services;

/// <summary>Parses a Moonraker printer/objects/query response into PrinterTelemetry, including the
/// Happy Hare mmu object mapped to AMS slots. Mirrors the macOS MoonrakerStatusParser.</summary>
public static class MoonrakerStatusParser
{
    public static PrinterTelemetry? Telemetry(byte[] data, PrinterTelemetry? previous = null)
    {
        JsonElement root;
        try { using var doc = JsonDocument.Parse(data); root = doc.RootElement.Clone(); }
        catch { return null; }

        JsonElement result = root.TryGetProperty("result", out var r) ? r : root;
        if (!result.TryGetProperty("status", out var status) || status.ValueKind != JsonValueKind.Object) return null;

        var t = previous?.Clone() ?? new PrinterTelemetry();

        if (Obj(status, "print_stats", out var printStats))
        {
            if (Str(printStats, "state") is { } state) t.State = MapState(state);
            if (Str(printStats, "filename") is { Length: > 0 } file) t.JobName = Path.GetFileName(file);
            if (Obj(printStats, "info", out var info))
            {
                if (Int(info, "current_layer") is { } cl) t.CurrentLayer = cl;
                if (Int(info, "total_layer") is { } tl) t.TotalLayers = tl;
            }
        }

        double? progress = null;
        if (Obj(status, "display_status", out var ds)) progress = Num(ds, "progress");
        if (progress is null && Obj(status, "virtual_sdcard", out var vs)) progress = Num(vs, "progress");
        if (progress is { } p) t.Progress = Math.Clamp((int)Math.Round(p * 100), 0, 100);

        if (Obj(status, "print_stats", out var ps) && Num(ps, "print_duration") is { } duration && duration > 0
            && progress is { } pr && pr > 0.01)
        {
            t.RemainingMinutes = (int)Math.Round(duration * (1 - pr) / pr / 60);
        }

        if (Obj(status, "extruder", out var extruder))
        {
            if (Num(extruder, "temperature") is { } temp) t.NozzleTemperature = temp;
            if (Num(extruder, "target") is { } target) t.NozzleTargetTemperature = target;
        }
        if (Obj(status, "heater_bed", out var bed))
        {
            if (Num(bed, "temperature") is { } temp) t.BedTemperature = temp;
            if (Num(bed, "target") is { } target) t.BedTargetTemperature = target;
        }
        if (ChamberTemperature(status) is { } chamber) t.ChamberTemperature = chamber;

        if (Obj(status, "mmu", out var mmu))
        {
            var slots = ParseMmu(mmu);
            if (slots.Count > 0) t.AmsSlots = slots;
        }

        t.LastUpdated = DateTime.Now;
        return t;
    }

    public static PrinterState MapState(string raw) => raw.ToLowerInvariant() switch
    {
        "printing" => PrinterState.Printing,
        "paused" => PrinterState.Paused,
        "complete" or "completed" => PrinterState.Finished,
        "error" => PrinterState.Error,
        "cancelled" or "canceled" or "standby" => PrinterState.Idle,
        _ => PrinterState.Idle
    };

    private static double? ChamberTemperature(JsonElement status)
    {
        foreach (var prop in status.EnumerateObject())
        {
            if (!prop.Name.ToLowerInvariant().Contains("chamber")) continue;
            if ((prop.Name.StartsWith("temperature_sensor") || prop.Name.StartsWith("heater_generic"))
                && prop.Value.ValueKind == JsonValueKind.Object && Num(prop.Value, "temperature") is { } temp)
                return temp;
        }
        return null;
    }

    private static List<AmsSlot> ParseMmu(JsonElement mmu)
    {
        if (mmu.TryGetProperty("enabled", out var enabled) && enabled.ValueKind == JsonValueKind.False) return new();
        if (Int(mmu, "num_gates") is not { } count || count <= 0) return new();

        var materials = Arr(mmu, "gate_material");
        var colors = Arr(mmu, "gate_color");
        var statuses = Arr(mmu, "gate_status");
        int current = Int(mmu, "gate") ?? -1;

        var slots = new List<AmsSlot>();
        for (int i = 0; i < count; i++)
        {
            int gateStatus = i < statuses.Count ? (IntValue(statuses[i]) ?? -1) : -1;
            string rawMaterial = i < materials.Count ? (StringValue(materials[i]) ?? "") : "";
            string material = gateStatus == 0 || rawMaterial.Length == 0 ? "—" : rawMaterial;
            string color = AmsColor(i < colors.Count ? StringValue(colors[i]) : null);
            slots.Add(new AmsSlot
            {
                Id = $"mmu-{i}",
                Label = $"T{i}",
                Material = material,
                ColorHex = color,
                RemainingPercent = null,
                IsActive = i == current,
                IsExternal = false
            });
        }
        return slots;
    }

    private static string AmsColor(string? raw)
    {
        if (string.IsNullOrEmpty(raw)) return "8E8E93FF";
        var value = raw.StartsWith('#') ? raw[1..] : raw;
        if (value.Length == 6) return (value + "FF").ToUpperInvariant();
        if (value.Length == 8) return value.ToUpperInvariant();
        return "8E8E93FF";
    }

    private static bool Obj(JsonElement parent, string key, out JsonElement value)
    {
        if (parent.ValueKind == JsonValueKind.Object && parent.TryGetProperty(key, out value) && value.ValueKind == JsonValueKind.Object)
            return true;
        value = default;
        return false;
    }

    private static List<JsonElement> Arr(JsonElement parent, string key)
    {
        if (parent.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.Array)
            return value.EnumerateArray().ToList();
        return new();
    }

    private static string? Str(JsonElement obj, string key)
        => obj.TryGetProperty(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static double? Num(JsonElement obj, string key)
        => obj.TryGetProperty(key, out var v) ? NumberValue(v) : null;

    private static int? Int(JsonElement obj, string key)
    {
        var n = Num(obj, key);
        return n.HasValue ? (int)n.Value : null;
    }

    private static double? NumberValue(JsonElement v)
    {
        if (v.ValueKind == JsonValueKind.Number && v.TryGetDouble(out var d)) return d;
        if (v.ValueKind == JsonValueKind.String && double.TryParse(v.GetString(), NumberStyles.Any, CultureInfo.InvariantCulture, out var ds)) return ds;
        return null;
    }

    private static string? StringValue(JsonElement v) => v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static int? IntValue(JsonElement v)
    {
        var n = NumberValue(v);
        return n.HasValue ? (int)n.Value : null;
    }
}

File: Sources/BambuBar/Services/MoonrakerStatusParser.swift
import Foundation

/// Parses a Moonraker `printer/objects/query` response into PrinterTelemetry, including the
/// Happy Hare `mmu` object mapped to AMS slots. Field names verified defensively; unknown or
/// missing values fall back to the previous telemetry.
enum MoonrakerStatusParser {
    static func telemetry(from data: Data, previous: PrinterTelemetry = .init()) -> PrinterTelemetry? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        let result = root["result"] as? [String: Any] ?? root
        guard let status = result["status"] as? [String: Any] else { return nil }

        var telemetry = previous

        if let printStats = status["print_stats"] as? [String: Any] {
            if let state = string(printStats["state"]) { telemetry.state = mapState(state) }
            if let name = string(printStats["filename"]), !name.isEmpty {
                telemetry.jobName = (name as NSString).lastPathComponent
            }
            if let info = printStats["info"] as? [String: Any] {
                if let current = integer(info["current_layer"]) { telemetry.currentLayer = current }
                if let total = integer(info["total_layer"]) { telemetry.totalLayers = total }
            }
        }

        // Progress: display_status is the slicer/M73 value; fall back to sd position.
        let progress = number((status["display_status"] as? [String: Any])?["progress"])
            ?? number((status["virtual_sdcard"] as? [String: Any])?["progress"])
        if let progress {
            telemetry.progress = min(max(Int((progress * 100).rounded()), 0), 100)
        }

        // ETA from elapsed print time and progress.
        if let printStats = status["print_stats"] as? [String: Any],
           let duration = number(printStats["print_duration"]), duration > 0,
           let progress, progress > 0.01 {
            let remainingSeconds = duration * (1 - progress) / progress
            telemetry.remainingMinutes = Int((remainingSeconds / 60).rounded())
        }

        if let extruder = status["extruder"] as? [String: Any] {
            if let temp = number(extruder["temperature"]) { telemetry.nozzleTemperature = temp }
            if let target = number(extruder["target"]) { telemetry.nozzleTargetTemperature = target }
        }
        if let bed = status["heater_bed"] as? [String: Any] {
            if let temp = number(bed["temperature"]) { telemetry.bedTemperature = temp }
            if let target = number(bed["target"]) { telemetry.bedTargetTemperature = target }
        }
        // Chamber: any temperature_sensor / heater_generic whose name mentions "chamber".
        if let chamber = chamberTemperature(in: status) { telemetry.chamberTemperature = chamber }

        if let mmu = status["mmu"] as? [String: Any] {
            let slots = parseMMU(mmu)
            if !slots.isEmpty { telemetry.amsSlots = slots }
        }

        telemetry.lastUpdated = Date()
        return telemetry
    }

    static func mapState(_ raw: String) -> PrinterState {
        switch raw.lowercased() {
        case "printing": .printing
        case "paused": .paused
        case "complete", "completed": .finished
        case "error": .error
        case "cancelled", "canceled", "standby": .idle
        default: .idle
        }
    }

    private static func chamberTemperature(in status: [String: Any]) -> Double? {
        for (key, value) in status where key.lowercased().contains("chamber") {
            if (key.hasPrefix("temperature_sensor") || key.hasPrefix("heater_generic")),
               let object = value as? [String: Any],
               let temp = number(object["temperature"]) {
                return temp
            }
        }
        return nil
    }

    private static func parseMMU(_ mmu: [String: Any]) -> [AMSSlot] {
        if let enabled = mmu["enabled"] as? Bool, !enabled { return [] }
        guard let count = integer(mmu["num_gates"]), count > 0 else { return [] }

        let materials = mmu["gate_material"] as? [Any] ?? []
        let colors = mmu["gate_color"] as? [Any] ?? []
        let statuses = mmu["gate_status"] as? [Any] ?? []
        let currentGate = integer(mmu["gate"]) ?? -1

        var slots: [AMSSlot] = []
        for index in 0..<count {
            let gateStatus = index < statuses.count ? (integer(statuses[index]) ?? -1) : -1
            let rawMaterial = index < materials.count ? (string(materials[index]) ?? "") : ""
            let material = gateStatus == 0 || rawMaterial.isEmpty ? "—" : rawMaterial
            let color = amsColor(index < colors.count ? string(colors[index]) : nil)
            slots.append(AMSSlot(
                id: "mmu-\(index)",
                label: "T\(index)",
                material: material,
                colorHex: color,
                remainingPercent: nil,
                isActive: index == currentGate,
                isExternal: false
            ))
        }
        return slots
    }

    /// Happy Hare gate colours are hex (6 chars) or a named colour; normalise to RRGGBBAA.
    private static func amsColor(_ raw: String?) -> String {
        guard var value = raw, !value.isEmpty else { return "8E8E93FF" }
        if value.hasPrefix("#") { value.removeFirst() }
        if value.count == 6 { return (value + "FF").uppercased() }
        if value.count == 8 { return value.uppercased() }
        return "8E8E93FF"
    }

    private static func string(_ value: Any?) -> String? { value as? String }

    private static func number(_ value: Any?) -> Double? {
        if let number = value as? NSNumber { return number.doubleValue }
        if let string = value as? String { return Double(string) }
        return nil
    }

    private static func integer(_ value: Any?) -> Int? { number(value).map(Int.init) }
}
